File: AE/Arrays/non_constructible_change.py
"""
- input: an array of positive integers representing the values of coins
- output: return the min amount of change (min sum of money) that you CANNOT create
- notes:
  - given coins can have any positive integer values and aren't necessarily unique (you can have multiple coins of the same value)
  - example: coins = [1, 2, 5], and the minimum amount of change you cannot create is 4
  - if not given any coins, the min amount of change you cannot create is 1
- Logic:
  - since looking for min change that cannot be made, sort the coins array
  - create a change variable that keeps track of all the change you can create; for example change = 3 means you can create change for a value of 1, 2, and 3
  - iterate through the coins array
    - if coin value is greater than change value + 1, return change value + 1 --> because the coin you're about to receive will allow you to make change for that coin value and any combination of the coin value + range of change value from 1 to change value, but not change value + 1
    = if not, add coin to change
  - at the end return change + 1 --> if coins array is empty than change value will be 0, thus we cannot make 0 + 1 = 1 amount of change
"""


def non_constructible_change(coins):
    coins.sort()

    change = 0

    for coin in coins:
        if coin > change + 1:
            return change + 1
        change += coin

    return change + 1

File: AE/Arrays/nonConstructibleChange.js
/*
Given an array of positive integers representing the values of coins in your possession, write a function that returns the minimum amount of change (the minimum sum of money) that you cannot create. The given coins can have any positive integer value and aren't necessarily unique (e.g. you can have multiple coins of the same value).

For example, if given [1,2,5], the minimum amount of change you cannot create is 4. You can make change for 1, 2, 3 (1+2), and 5, but not 4. If no given coins, the minimum amount of change you can't create is 1

Sample Input:
coins = [5,7,1,1,2,3,22]

Sample Output:
20

*/

const nonConstructibleChangeNaive = (coins) => {
  
  //in search for minimum amount of change that cannot be created, so sort array
  coins.sort((a,b) => a - b);
  //track the change created
  let changeCreated = 0;

  for (let i = 0; i < coins.length; i++) {
    const coin = coins[i];

    //if the coin value is greater than the change created so far + 1
    //then we can create change for every value from 1 to changeCreated and the coin value, but not changeCreated + 1
    if (coin > changeCreated + 1) return changeCreated + 1;

    changeCreated += coin;
  } 
  return changeCreated + 1;
}

// Time Complexity: O(NlogN);
// Space Complexity: O(1);